### ISSUE TITLE
Checkout Internal Server Error

### ISSUE THREAD

**Issue by gofas** [woothemes/woocommerce#9279]
Hey team,
i have problems in the checkout page, apparently jquery not send value of fields to ajax. 

After click in "Go to PayPal" the "Internal Server Error" message is displayed in top of page, and in the Order Panel register a empty order.

![pedidos_wordpress](https://cloud.githubusercontent.com/assets/7539811/10263845/2d29e3fc-69d2-11e5-9469-53e4d8cf3d44.png)

Even after disable all plugins and test with the standard template the error persists. 
I use nginx in ubuntu 14.4.x, my first suspicion it was configuration of my  /etc/nginx/sites-available/site.com.

Would if I share my setup and my log you guys can help me find the way to the solution?

Thank you for the incredible work you are doing!!

Those are my logs and settings:

Chrome console error:
Failed to load resource: the server responded with a status of 500 (Internal Server Error)
jquery.js:5 POST http://site.com/finalizar-compra/?wc-ajax=checkout 500 (Internal Server Error)send @ jquery.js:5m.extend.ajax @ jquery.js:5b.submit @ checkout.min.js:1m.event.dispatch @ jquery.js:4r.handle @ jquery.js:4

My Woocommerce system status:
https://gist.github.com/gofas/bb7251b41a2cd8560c3a

My nginx generated site.com.log:
https://gist.github.com/gofas/3f39def675431ad17581

My wp-content/debug.log:
https://gist.github.com/gofas/f0e4df799b88fb966bd3

My complete server config:
https://gist.github.com/gofas/d50672cf6660ca030102
https://gist.github.com/gofas/47ab70a3f7aa10c660c3
https://gist.github.com/gofas/634c4dfd6679cf055205
https://gist.github.com/gofas/905a702e4305d2fc6f80
https://gist.github.com/gofas/8f23e9cff2fcb488edaf

**Comment by mikejolley** [comment]
You will get this error if the order is not created/returned properly. Looking at your logs thats what is suggested, although I cannot see 'why'.

You will need to disable all plugins to narrow this one down, and retest. You may also want to reinstall WC to ensure all files are up to date.

**Comment by gofas** [comment]
Sorry, I forgot to mention that this site runs in a multisite network (where I have two more stores with the same error).
I disabled previously all plugins of the network and of the subsite to test and collect the logs, also tested in twenty fifteen and storefront (parent) themes.

It also reinstalled WooCommerce (various earlier versions ), the core of wp, plugins and themes, all.

My log:
'[06-Oct-2015 01:44:14 UTC] PHP Fatal error:  Call to a member function add_product() on a non-object in /var/www/site.com/htdocs/wp-content/plugins/woocommerce/includes/class-wc-checkout.php on line 217'
full:
https://gist.github.com/gofas/68ac5cff0efa62d7681b

**Comment by mikejolley** [comment]
Yup, from that line, no $order exists. In that class, does commenting out:

`$order_id = absint( WC()->session->order_awaiting_payment );`

fix it?

**Comment by gofas** [comment]
Hello Mike!
Sorry if my understanding is limited, i ought comment this line?
I did it, but it not work :( ...

But on the other hand, i sucess run the checkout after override woocommerce front end js scripts for scripts found on version 2.2.9 of wc.

1 - I copy `/woocommerce/assets/js/frontend/` to `/my-child-theme/woocommerce/assets/js/frontend/`
2 - I add this action to override scripts:
`// Load woocommerce frontend js v 2.2.9
add_action('wp_enqueue_scripts', 'override_woo_frontend_scripts');
function override_woo_frontend_scripts() {
    wp_deregister_script('wc-checkout');
    wp_enqueue_script('wc-checkout', get_stylesheet_directory_uri() . '/woocommerce/js/checkout.js', array('jquery', 'woocommerce', 'wc-country-select', 'wc-address-i18n'), null, true);
}
// End Load woocommerce frontend js v 2.2.9
`
It work for me, the website is http://www.oficioartebambu.com.br/.

Thanks for your time!

**Comment by mikejolley** [comment]
Are you actually able to access the endpoint http://site.com/finalizar-compra/?wc-ajax=checkout ?

Running a security plugin?

**Comment by gofas** [comment]
It is accessible but returns an error, see for yourself:
http://www.oficioartebambu.com.br/finalizar-compra/?wc-ajax=checkout

Yes, i running Wordfence in network.

**Comment by mikejolley** [comment]
Looks like a line of white space before the JSON - that will be an issue.

**Comment by gofas** [comment]
Yes, i figured it out... How can i fix it?

**Comment by mikejolley** [comment]
You need to narrow down the cause. Plugin, theme, or white space in a config file for instance.

**Comment by gofas** [comment]
wow, that's a start, i will investigate...
The same issue is going on http://www.oficioartebambu.com.br/feed/,
the wp-blog-header.php might be the answer, but it is ok, no white space :(

**Comment by gofas** [comment]
now I'm extremely confused when I turn off WooCommerce the error is corrected, including feeds

**Comment by mikejolley** [comment]
Maybe its a plugin *for* woocommerce?

**Comment by gofas** [comment]
It fix blank line in /?wc-ajax=checkout:

`define('WP_DEBUG', false);
error_reporting(0);
@ini_set('display_errors', 0);` 

but not solve the "Internal Server Error"

**Comment by mikejolley** [comment]
Thats just hiding the line. You have a problem with something on your site and need to find it rather than hide it. 

- so check plugins again
- check logs again
- if orders are not created it's usually due to a plugin, or no more space on the server

I'm closing this issue because this isn't a core WC bug (it's something in your install), and we are compatible with nginx. My own stores run on nginx for example without issue.

**Comment by goblindegook** [comment]
FWIW, we had the same issue and discovered it was caused by the `wc_get_order()` function returning `false` immediately after creating the order.

Because we had no issues locally and our environment is supported by MariaDB w/ Galera Cluster, we dug deeper and experimented with the transaction queries.

Turns out WooCommerce is relying on dirty reads (`READ-UNCOMMITTED` isolation level) to function properly.  Unless you enable that, any attempts to get the just-created order from the database (using `wc_get_order()`, etc. will fail).

If you're using a clustered database like us, you have two ways out of this: you can either allow dirty reads, or you can disable transactions completely. Note that you can only have dirty reads on master-slave setups, if you're on a multi-master Galera Cluster, you're forced to use `REPEATABLE-READ`.

The following snippet should allow you to disable transactions within WordPress:

```php
add_filter( 'query', function( $query ) {
	$pattern = '/^\s*(START TRANSACTION|COMMIT|ROLLBACK)/i';
	return preg_match( $pattern, $query ) ? '' : $query;
}
```

**Comment by gofas** [comment]
Wow, it's fine work @goblindegook!!!

It makes me more relaxed knowing that the problem is not on my wp installation. That is not a compatibility failure to be studied @mikejolley?

Thank for all dedication guys!

**Comment by mikejolley** [comment]
@goblindegook @gofas Any chance you can look into a solution that doesn't disable translations?

I don't think why you're affected; if you look here you'll see that only the order ID is returned AFTER the commit: https://github.com/woothemes/woocommerce/blob/master/includes/class-wc-checkout.php#L335

The order is then retrieved later from said ID: https://github.com/woothemes/woocommerce/blob/master/includes/class-wc-checkout.php#L657

Where is the dirty read?

**Comment by goblindegook** [comment]
@mikejolley: the dirty read is at the very end of the `wc_create_order()` core function, which is called during the checkout transaction:

* https://github.com/woothemes/woocommerce/blob/master/includes/class-wc-checkout.php#L205
* https://github.com/woothemes/woocommerce/blob/master/includes/wc-core-functions.php#L120

Maybe try to `SET TRANSACTION ISOLATION LEVEL READ UNCOMMITTED` to see if you get an error? You could then either throw an exception or continue without transactions.

Alternatively, you may be able to create a `WC_Order` instance without having to go into the database, but that could still cause issues with hooks that do database reads.

**Comment by goblindegook** [comment]
Since reads are (AFAICT) index-based, `READ COMMITTED` may be a better option. It's still not supported in multi-master environments, though, so you should definitely check for errors after trying to set this isolation level.

**Comment by raisiqueira** [comment]
@gofas desativa todos os teus plugins, deixa somente o WC. Tive o mesmo problema aqui, um plugin estava gerando conflito com o jquery.

**Comment by mikejolley** [comment]
@goblindegook I don't get errors on my install with any setting. If you place:

`
$wpdb->query( 'SET SESSION TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;' );
`

Above the start transaction line, does it solve your error? Is that on Maria DB?

**Comment by goblindegook** [comment]
We're on a Galera Cluster multi-master setup, which supports `REPEATABLE READ` only.

**Comment by gokhandmr** [comment]
hi my same problems. Please help me!! :( I could not figure it's been weeks.

my debug.log inside;

[01-Sep-2016 20:52:12 UTC] PHP Notice:  load_plugin_textdomain was called with an argument that is <strong>deprecated</strong> since version 2.7.0 with no alternative available. in /home/admin/web/proteintozu.com.tr/public_html/wp-includes/functions.php on line 3892
[01-Sep-2016 20:52:14 UTC] PHP Notice:  load_plugin_textdomain was called with an argument that is <strong>deprecated</strong> since version 2.7.0 with no alternative available. in /home/admin/web/proteintozu.com.tr/public_html/wp-includes/functions.php on line 3892
[01-Sep-2016 20:52:18 UTC] PHP Notice:  load_plugin_textdomain was called with an argument that is <strong>deprecated</strong> since version 2.7.0 with no alternative available. in /home/admin/web/proteintozu.com.tr/public_html/wp-includes/functions.php on line 3892
[01-Sep-2016 20:52:21 UTC] PHP Notice:  load_plugin_textdomain was called with an argument that is <strong>deprecated</strong> since version 2.7.0 with no alternative available. in /home/admin/web/proteintozu.com.tr/public_html/wp-includes/functions.php on line 3892
[01-Sep-2016 23:52:22 Europe/istanbul] PHP Notice:  Undefined index: 00 in /home/admin/web/proteintozu.com.tr/public_html/wp-includes/class-wp-locale.php on line 312
[01-Sep-2016 20:52:23 UTC] PHP Notice:  load_plugin_textdomain was called with an argument that is <strong>deprecated</strong> since version 2.7.0 with no alternative available. in /home/admin/web/proteintozu.com.tr/public_html/wp-includes/functions.php on line 3892
[01-Sep-2016 20:52:26 UTC] PHP Notice:  load_plugin_textdomain was called with an argument that is <strong>deprecated</strong> since version 2.7.0 with no alternative available. in /home/admin/web/proteintozu.com.tr/public_html/wp-includes/functions.php on line 3892
[01-Sep-2016 23:52:27 Europe/istanbul] PHP Parse error:  syntax error, unexpected end of file in /home/admin/web/proteintozu.com.tr/public_html/wp-content/plugins/woocommerce/templates/emails/email-styles.php on line 179

My Functions.php   3892 line inside:

		trigger_error( sprintf( __('%1$s was called with an argument that is <strong>deprecated</strong> since version %2$s with no alternative available.'), $function, $version ) );
		} else {
			if ( ! is_null( $message ) )
				trigger_error( sprintf( '%1$s was called with an argument that is <strong>deprecated</strong> since version %2$s! %3$s', $function, $version, $message ) );
			else
				trigger_error( sprintf( '%1$s was called with an argument that is <strong>deprecated</strong> since version %2$s with no alternative available.', $function, $version ) );
		}
	}
}

email-styles.php on line 179 inside;

there is only this : <?php

  not my code information. I do not know what to do. help please :(

**Comment by gokhandmr** [comment]
Hi, nobody answered but i fixed it. Maybe will another people help. I did my ubuntu server and mysql server update. Sorry for my english :) have a nice day!

**Comment by mauriciogofas** [comment]
Unfortunately is still necessary disable transactions on mariadb cluster servers and multisite installations with multi DB.

**Comment by joelpeterson** [comment]
Hey guys... I'm suspecting I'm running into the same issue and want to try to add the snipit...

Dumb question.... what file do I add the code too?

add_filter( 'query', 'fix_ajax_query');
function fix_ajax_query( $query ) {
$pattern = '/^\s*(START TRANSACTION|COMMIT|ROLLBACK)/i';
return preg_match( $pattern, $query ) ? '' : $query;
}

**Comment by baseconversion** [comment]
Joel: In a custom plugin or your theme's functions.php files. If you want it to apply only to your theme then add it there. If you want it to apply to more than just one theme then add it to a custom plugin.

**Comment by girlskull** [comment]
I don't know if this helps anyone but I had the same problem with my website at checkout it would throw the error and then all these uncompleted non paid orders would end up in my woocommerce message box.  What worked for me was disabling a plugin called emember woocommerce addon.  That fixed it right away.  Hope this helps.

**Comment by AntonioCorrenti** [comment]
Hi,
I just recently changed machine and had this problem.
I solved by installing some missing php extensions.
`sudo apt-get install php-curl
sudo apt-get install php-mbstring
sudo apt-get install php-xml
sudo service apache2 restart
`

**Comment by tsisibedy** [comment]
*baro lima omo laray woocommerce me lakalzoafha pluging
**ligaofha liamora!

**Comment by amirmagictv** [comment]
Hello. I had the same issue and was trying different things, however, what resolved my issue is switching PHP version to 7.2. 
a2dismod php7.1
a2enmod php7.2
service apache2 restart
